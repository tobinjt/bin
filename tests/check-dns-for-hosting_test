#!/bin/bash

set -e -f -u -o pipefail

# shellcheck disable=SC1091
source ../check-dns-for-hosting

SECONDS_TO_SLEEP_BETWEEN_ATTEMPTS=0
export SECONDS_TO_SLEEP_BETWEEN_ATTEMPTS

# Set up fake host records.
host_records="$(mktemp -t host_records.XXXXXXXXXXXXXXXXX)"
host() {
  local domain="$3"
  grep "^${domain} " "${host_records}"
}
cat > "${host_records}" <<HOST_RECORDS
# domain1.ie will succeed.
domain1.ie has address 88.99.86.190
domain1.ie has IPv6 address 2a01:4f8:c17:7156::2
www.domain1.ie has address 88.99.86.190
www.domain1.ie has IPv6 address 2a01:4f8:c17:7156::2
domain1.ie mail is handled by 1 aspmx.l.google.com.
domain1.ie mail is handled by 10 aspmx2.googlemail.com.
domain1.ie mail is handled by 10 aspmx3.googlemail.com.
domain1.ie mail is handled by 5 alt1.aspmx.l.google.com.
domain1.ie mail is handled by 5 alt2.aspmx.l.google.com.

# dev.domain2.ie will succeed, without subdomains being checked.
dev.domain2.ie has address 88.99.86.190
dev.domain2.ie has IPv6 address 2a01:4f8:c17:7156::2

# domain3.ie will fail because there are no records.
HOST_RECORDS

stdout="$(mktemp -t stdout.XXXXXXXXXXXXXXXXX)"
stderr="$(mktemp -t stderr.XXXXXXXXXXXXXXXXX)"
{ main "domain1.ie" || printf "Unexpected failure for domain1.ie\\n" >&2;
  main "dev.domain2.ie" || printf "Unexpected failure for domain1.ie\\n" >&2;
  main "domain3.ie" || exit_status="$?";
  if [[ "${exit_status}" -eq 0 ]]; then
    printf "Unexpected success for domain3.ie\\n" >&2;
  fi;
} > "${stdout}" 2> "${stderr}"

# Test that retries work.
# First time host is run output nothing, set up output for the second time.
host() {
  # Second time it prints the A record and sets up for the third time.
  host() {
    cat <<HOST_RECORDS
sub.domain4.ie has address 88.99.86.190
HOST_RECORDS

    # Third time it outputs the AAAA record and sets up to fail from now on.
    host() {
      cat <<HOST_RECORDS
sub.domain4.ie has IPv6 address 2a01:4f8:c17:7156::2
HOST_RECORDS

      # Fourth and subsequent times it should not be called.
      host() {
        printf "Host should not have been called a fourth time\\n" >&2
        return 1
      }
    }
  }
}

{ main "sub.domain4.ie" || printf "Unexpected failure for domain4.ie\\n" >&2;
} >> "${stdout}" 2>> "${stderr}"

# Clean up variable output.
sed -i .bak -e 's/host output in.*/host output in tmpfile/' "${stderr}"
diff -u "check-dns-for-hosting_test.stdout" "${stdout}"
diff -u "check-dns-for-hosting_test.stderr" "${stderr}"
rm -f "${stdout}" "${stderr}"
