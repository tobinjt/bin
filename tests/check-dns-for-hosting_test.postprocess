#!/bin/bash

set -e -f -u -o pipefail

stderr="$2"
# Clean up variable output.
sed -i.bak -e 's/host output in.*/host output in tmpfile/' "${stderr}"
rm -f "${stderr}.bak"
