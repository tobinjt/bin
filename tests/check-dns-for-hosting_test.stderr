Bad A address: pattern '^domain3.ie has address 168.119.99.114$' not found
Bad AAAA address: pattern '^domain3.ie has IPv6 address 2a01:4f8:c010:21a::1$' not found
host output in tmpfile
Bad A address: pattern '^www.domain3.ie has address 168.119.99.114$' not found
Bad AAAA address: pattern '^www.domain3.ie has IPv6 address 2a01:4f8:c010:21a::1$' not found
host output in tmpfile
Bad MX address: pattern '^domain3.ie mail is handled by 1 aspmx.l.google.com.$' not found
Bad MX address: pattern '^domain3.ie mail is handled by 10 aspmx2.googlemail.com.$' not found
Bad MX address: pattern '^domain3.ie mail is handled by 10 aspmx3.googlemail.com.$' not found
Bad MX address: pattern '^domain3.ie mail is handled by 5 alt1.aspmx.l.google.com.$' not found
Bad MX address: pattern '^domain3.ie mail is handled by 5 alt2.aspmx.l.google.com.$' not found
host output in tmpfile
